File: src/main/kotlin/cz/smartbrains/qesu/module/common/service/CodeListService.kt
package cz.smartbrains.qesu.module.common.service

import cz.smartbrains.qesu.module.common.dto.CodeListDto
import cz.smartbrains.qesu.module.common.entity.CodeList

interface CodeListService<E : CodeList, T : CodeListDto> : OrdeableEntityService<T> {
    fun findAll(): List<T>
    fun create(codeListDto: T): T
    fun update(codeListDto: T): T
    fun delete(id: Long)
}

File: src/main/client/src/modules/common/services/codeListService.ts
import CodeList from '@/modules/common/domain/codeList'
import EntityResource from '@/modules/common/services/entityResource'

export default class CodeListService<T extends CodeList> {
  resource: EntityResource

  constructor (repository: EntityResource) {
    this.resource = repository
  }

  async findAll () {
    const response = await this.resource.query()
    if (response.ok) {
      return response.data.map((itemData: any) => this.createItem(itemData))
    } else {
      return null
    }
  }

  async create (item: T) {
    const response = await this.resource.save({}, item)
    if (response.ok) {
      return this.createItem(response.data)
    } else {
      return null
    }
  }

  async update (item: T) {
    const response = await this.resource.update({}, item)
    if (response.ok) {
      return this.createItem(response.data)
    } else {
      return null
    }
  }

  async delete (id: bigint) {
    return this.resource.delete({ id })
  }

  async updatePosition (id: bigint, position: bigint) {
    const response = await this.resource.updatePosition({ id, position }, {})
    return response.ok
  }

  createItem (data: any) {
    return new CodeList(data)
  }
}
